File: spring/src/main/java/org/jonas/test/spring/conf/bean/abs/TestApp.java
package org.jonas.test.spring.conf.bean.abs;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.CommandLineRunner;
import org.springframework.boot.SpringApplication;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.boot.context.properties.ConfigurationProperties;
import org.springframework.boot.context.properties.EnableConfigurationProperties;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

@SpringBootApplication
public class TestApp implements CommandLineRunner {

    public static void main(String[] args) {
        SpringApplication.run(TestApp.class, args);
    }

    @Autowired
    Data2 data2;

    @Override
    public void run(String... args) throws Exception {
        System.out.println(data2.getName());
    }
}

@Configuration
@EnableConfigurationProperties(Data1.class)
class Conf1 {
}

abstract class Data1 {
    abstract String getName();
}

@ConfigurationProperties
class Data2 extends Data1 {
    @Override
    String getName() {
        return "Hello Spring!";
    }
}

File: spring/src/main/java/org/jonas/test/spring/conf/bean/call/TestApp.kt
package org.jonas.test.spring.conf.bean.call

import org.springframework.boot.CommandLineRunner
import org.springframework.boot.SpringApplication
import org.springframework.boot.autoconfigure.SpringBootApplication
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.core.io.support.ResourcePatternResolver

fun main(args: Array<String>) {
    SpringApplication.run(TestApp::class.java, *args)
}

@SpringBootApplication
class TestApp(val data1: Data1, val data2: Data2) : CommandLineRunner {
    override fun run(vararg args: String?) {
        println(data1.data.name)
        println(data2.data1 == data1)
    }
}

@Configuration
class Conf1 {
    @Bean
    fun data1(data31: Data3): Data1 {
        println("create data 1 !!")
        return Data1(data31)
    }

    @Bean
    fun data2(data32: Data3): Data2 {
        println(data32.name)
        return Data2(data1(data32))
    }
}

@Configuration
class Conf2 {
    @Bean fun data31() = Data3("One!")
    @Bean fun data32() = Data3("Two!")
}

class Data1(val data: Data3)

class Data2(val data1: Data1)

class Data3(val name: String)